lambda/trmc.ml

```diff
@@ -0,0 +1,515 @@
+    the offset is an Ident.t), and maybe in the future statically-known
+    offsets (where the offset is an integer).
+*)
+
+let add_dst_params ({var; offset} : Ident.t destination) params =
+  (var, Pgenval) :: (offset, Pintval) :: params
+
+let add_dst_args ({var; offset} : offset destination) args =
+  Lvar var :: offset :: args
+
+let assign_to_dst loc {var; offset} lam =
+  Lprim(Psetfield_computed(Pointer, Heap_initialization),
+        [Lvar var; offset; lam], loc)
+
+(** TRMC transformation requires information flows in two opposite
+    directions: the information of which callsites can be rewritten in
+    destination-passing-style flows from the leaves of the code to the
+    root, and the information on whether we remain in tail-position
+    flows from the root to the leaves -- and also the knowledge of
+    which version of the function we currently want to generate, the
+    direct version or a destination-passing-style version.
+
+    To clarify this double flow of information, we split the TRMC
+    compilation in two steps:
+
+    1. A function [trmc t] that takes a term and processes it from
+    leaves to root; it produces a "code choice", a piece of data of
+    type [choice], that contains information on how to transform the
+    input term [t] *parameterized* over the (still missing) contextual
+    information.
+
+    2. Code-production operators that have contextual information
+    to transform a "code choice" into the final code.
+*)
+
+type choice =
+  | Return of return
+  (** [Return t] means that there are no TRMC opportunities in the subterm [t]:
+      no matter which context we are in,
+      we should evaluate [t] and "return" it. *)
+  | Set of settable
+  (** [Set t] represents a piece of code that does contain
+      TRMC opportunities: if the context allows, we can write parts of
+      it in destination-passing-style to turn non-tail calls into tail
+      calls. See the type [settable] below. *)
+
+and return = lambda
+
+and settable = {
+  dps: offset destination -> lambda;
+  direct: unit -> return;
+}
+(**
+   A [{dps; direct}] record a code that may be written in destination-passing style
+   if its usage context allows it. More precisely:
+
+   - If the surrounding context is already in destination-passing
+     style, it has a destination available, we should produce the
+     code in [dps] -- a function parametrized over the destination.
+
+  - If the surrounding context is in direct style (no destination
+    is available), we should produce the fallback code from
+    [direct].
+
+    (Note: [direct] is also a function (on [unit]) to ensure that any
+    effects performed during code production will only happen once we
+    do know that we want to produce the direct-style code.)
+ *)
+
+(** Finds the first [settable] element in a list [choices];
+    - if it exists, it gives a triple
+      [(rev_returns, settable, tail_choices)] such that
```

Comment:
Nit: this should be `[{rev_returns, settable, tail_choices}]`

Reply:
fixed, thanks.